src/rules-generator.ts

```diff
@@ -239,15 +226,61 @@ export class Generator {
 
     this.symbols.registerFunction('@RegExp', ['s'],
                                   ast.builtin(this.makeRegExp.bind(this)));
+
+    let map = this.symbols.registerSchema('Map', ast.typeType('Any'), undefined, undefined,
+                                          ['Key', 'Value']);
+    map.getValidator = this.getMapValidator.bind(this);
+  }
+
+  // type Map<Key, Value> => {
+  //   $key: {
+  //     '.validate': $key instanceof Key and this instanceof Value;
+  // }
+  // Key must derive from String
+  getMapValidator(params: ast.Exp[]): Validator {
+    let keyType = <ast.ExpSimpleType> params['Key'];
+    let valueType = <ast.ExpType> params['Value'];
+    if (keyType.type !== 'type' || !this.symbols.isDerivedFrom(keyType, 'String')) {
+      throw new Error(errors.invalidMapKey + "  (" + decodeExpression(keyType) + " does not)");
+    }
+
+    let validator = <Validator> {};
```

Comment:
I am quite confused over Validator and extendValidator. The comment on Validator doesn't help me figure out what we are doing here. Is extend doing a boolean & ?